video: common: prevent divide-by-zero in video_closest_frmival_stepwi…

sudarsan-22 opened the pull request:
…se()

Adds a check to ensure 'step' is not zero before using it in division, to prevent potential undefined behavior.

CID: 444378

sudarsan-22 committed:
video: common: prevent divide-by-zero video_closest_frmival_stepwise()

Adds an assertion and runtime log to ensure 'step' is not zero before
using it in division,preventing undefined behavior without modifying
 the public API.

Avoids a signature change (void → int) to preserve API stability for
Zephyr 4.2.A more complete solution can be proposed for 4.3.

CID: 444378

Signed-off-by: sudarsan N <sudarsansamy2002@gmail.com>

sudarsan-22 commented:
Fixes: #84783

josuah commented on drivers/video/video_common.c:
```diff
@@ -130,19 +130,39 @@ void video_closest_frmival_stepwise(const struct video_frmival_stepwise *stepwis
-	match->numerator = min + DIV_ROUND_CLOSEST(goal - min, step) * step;
-	match->denominator = stepwise->min.denominator * stepwise->max.denominator *
-			     stepwise->step.denominator * desired->denominator;
+	/* Prevent division by zero */
+	if (step == 0U) {
+		match->numerator = goal;
+		match->denominator = 1;
+		return;
+	}
```
When the step is zero, then this function should not be called as it works only on "stepwise" data with a non-zero step...

An error handling strategy seems more suited IMHO. For instance: `match->numerator = 0;`. Or `return -EINVAL` (would need to change the function type).

josuah reviewed (commented):
I think it is better to wait that https://github.com/zephyrproject-rtos/zephyr/pull/85697 is done, as this function will be slightly different as an outcome, so I propose pausing this until https://github.com/zephyrproject-rtos/zephyr/pull/85697 is completed.

Thank you very much for the help!

josuah reviewed (changes requested):
I meant to request changes, not approve, as there are still things we need to improve on this PR.

sudarsan-22 committed:
video: common: fix formatting for heap define to fit CI line length

Wrap K_HEAP_DEFINE to fit within 100-character CI limit and apply
consistent spacing.This is a formatting-only change.

Signed-off-by: sudarsan N <sudarsansamy2002@gmail.com>

sudarsan-22 commented:
> I meant to request changes, not approve, as there are still things we need to improve on this PR.

No worries, thanks for pointing it out! I’ll work on the improvements and update the PR,

josuah reviewed (dismissed):
Thank you for waiting this long and thank you for the fix!

danieldegrasse commented on include/zephyr/drivers/video.h:
```diff
@@ -818,8 +818,9 @@ static inline uint64_t video_frmival_nsec(const struct video_frmival *frmival)
+ * @return 0 on success, -EINVAL if step is 0 or input invalid
  */
-void video_closest_frmival_stepwise(const struct video_frmival_stepwise *stepwise,
+int  video_closest_frmival_stepwise(const struct video_frmival_stepwise *stepwise,
```
Little concerned by this API change in a public header, even if it is minor. Is it possible to fix this without modifying the API for this release? Or is this something we have to get done now?

josuah replied:
A division by zero would only happen when the user requests infinite FPS or zero FPS.

- If the caller does parameter validation then it might be an `__ASSERT_NO_MSG()` I guess.
- If the caller relies on it for parameter validation then it might need the `return` to check.

Glad to know the most "zephyr way" to do it to keep APIs uniform across areas.

danieldegrasse replied:
I am ok with `__ASSERT_NO_MSG`- also ok with logging an error and returning, would either of these work? To be clear I'm totally ok with this API changing in 4.3- just want to keep things stable for this release since we are already in the RC phase

josuah replied:
No pressure in this case, so may not be worth the risks of rushing it in 4.2.
Thank you for the vigilance!

josuah reviewed (approved):
Thank you, it looks like the best that can be done without changing the API.

avolmat-st commented on drivers/video/video_common.c:
```diff
@@ -136,6 +136,11 @@ void video_closest_frmival_stepwise(const struct video_frmival_stepwise *stepwis
 	step *= stepwise->min.denominator * stepwise->max.denominator * desired->denominator;
 	goal *= stepwise->min.denominator * stepwise->max.denominator * stepwise->step.denominator;
 
+	__ASSERT_NO_MSG(step != 0U);
+	/* Prevent division by zero */
+	if (step == 0U) {
+		return;
+	}
```
Agree, that wouldn't even reach the if.

josuah replied:
@sudarsan-22 yet another change if you do not mind :pray:

ngphibang replied:
Well, I think the driver will never let `uint64_t step = stepwise->step.numerator;` to be 0 (step = 0 makes no sense). So this will almost never happen except that the user deliberately or unintentionnally does that (pass a 0 step to the function).

Back to the function itself, this is a function that can always return something "closest" to what we gave to it (no success or failure). So, in this case, match is not computed (as we cannot step), all values are 0, or we can set them -1, as a "closest" value ?

josuah replied:
It seems like this could need API changes after all. As well as a bit of time for discussion.
Did anyone need this in 4.2?

ngphibang replied:
`__ASSERT_NO_MSG` is also an option. I don't know why we just avoid to use it in this case while it is just used in some lines above in the function and many other places ?

josuah replied:
I will following-up in https://github.com/zephyrproject-rtos/zephyr/issues/93024
Very glad to bring more checks to the API and constantly reduce efforts driver-side!
